flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AbstractEpochManager.java

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+import javax.annotation.Nonnull;
+
+import java.util.LinkedList;
+
+/**
+ * Epoch manager segments inputs into distinct epochs, marked by the arrival of non-records(e.g.
+ * watermark, record attributes). Records are assigned to a unique epoch based on their arrival,
+ * records within an epoch are allowed to be parallelized, while the non-record of an epoch can only
+ * be executed when all records in this epoch have finished.
+ *
+ * <p>For more details please refer to FLIP-425.
+ */
+public abstract class AbstractEpochManager {
+
+    /**
+     * This enum defines whether parallel execution between epochs is allowed. We should keep this
+     * internal and away from API module for now, until we could see the concrete need for {@link
+     * #PARALLEL_BETWEEN_EPOCH} from average users.
+     */
+    public enum ParallelMode {
+        /**
+         * Subsequent epochs must wait until the previous epoch is completed before they can start.
+         */
+        SERIAL_BETWEEN_EPOCH,
+        /**
+         * Subsequent epochs can begin execution even if the previous epoch has not yet completed.
+         * Usually performs better than {@link #SERIAL_BETWEEN_EPOCH}.
+         */
+        PARALLEL_BETWEEN_EPOCH
+    }
+
+    /** The output queue to hold ongoing epochs. */
+    protected LinkedList<Epoch> outputQueue;
+
+    public AbstractEpochManager() {
+        this.outputQueue = new LinkedList<>();
+        // preset an empty epoch, the epoch action will be updated when non-record is received.
+        this.outputQueue.add(new Epoch(0, () -> {}));
+    }
+
+    /**
+     * Add a record to the current epoch and return the current open epoch, the epoch will be
+     * associated with the {@link RecordContext} of this record. Must be invoked within task thread.
+     *
+     * @return the current open epoch.
+     */
+    public Epoch onRecord() {
+        Epoch lastEpoch = outputQueue.get(outputQueue.size() - 1);
+        lastEpoch.ongoingRecordCount++;
+        return lastEpoch;
+    }
+
+    /**
+     * Add a non-record to the current epoch, close current epoch and open a new epoch. Must be
+     * invoked within task thread.
+     *
+     * @param action the action associated with this non-record.
+     */
+    public abstract void onNonRecord(Runnable action);
+
+    /**
+     * Complete one record in the specific epoch. Must be invoked within task thread.
+     *
+     * @param epoch the specific epoch
+     */
+    public void completeOneRecord(Epoch epoch) {
+        epoch.ongoingRecordCount--;
+        // If this epoch has been closed before and all records in
+        //  is epoch have finished, the epoch will be removed from the output queue.
+        if (epoch.tryFinish() && outputQueue.size() > 0) {
+            if (epoch == outputQueue.getFirst()) {
+                outputQueue.remove(0);
```

Comment:
My suggestion: use `outputQueue.pop();` here. And there should be a
```
while (outputQueue.peek().isFinished()) {
    outputQueue.pop();
}
```
Right?

Reply:
right👍, I ignored the case that later epochs were completed first.